BWA on Spark experiment

tomwhite opened the pull request:
This is an experiment to see if it's possible to run BWA-MEM on Spark. (Please don't merge.) The basic idea is that it uses JNI to call BWA-MEM's align function to align a batch of read pairs in one go. I think it should complement the work that @SHuang-Broad has been doing in #1701. It would be great to get your (and @akiezun's) feedback on the direction here.

A few comments
* Building the native libraries is not integrated, and it's not using the Apache 2 licensed code. I think this could use some of the changes in #1701.
* The ref is assumed to be on the local FS for the moment - it should really be loaded from HDFS. Also, the output is a single SAM file on the local FS, not a sharded BAM as for the rest of the GATK Spark tools.
* It is assumed that read pairs are interleaved and reads in a pair are placed in the same split (by setting `hadoopbam.bam.keep-paired-reads-together`). However, that property only works for queryname sorted BAMs, which isn't the case here, so we need to relax that requirement in Hadoop-BAM.
* I haven't tried this on large inputs, so I don't know how well it performs.

To run, I used the following on a cluster

```
./gatk-launch BwaSpark \
  --ref /home/tom/workspace/jbwa/test/ref.fa \
  --input hdfs:///user/$USER/bwa/R.bam \
  --output /tmp/bwa.sam \
  -- \
  --sparkRunner SPARK --sparkMaster yarn-client \
  --driver-memory 3G \
  --num-executors 1 \
  --executor-cores 1 \
  --executor-memory 3G \
  --archives jbwa-native.tar#jbwa-native \
  --conf 'spark.executor.extraLibraryPath=jbwa-native'
```

The interesting bit is the use of Spark's `--archives` flag to copy a tarball of native libraries (which I built manually) to every executor, and unpacks it in the working directory. Then `spark.executor.extraLibraryPath` is set to add that path to the library path of the executor. This means that you don't have to rely on the native libraries being installed on every node in the cluster.

akiezun commented:
it does not compile. Is somethng missing?

tomwhite commented:
That's right. It's because I haven't added anything to gradle to build the native libraries. I suspect I could do that by incorporating that part of #1701. Should I go ahead and do that or wait for it to be merged?

SHuang-Broad commented:
@tomwhite , #1701 is not a JNI binding (I'm working on providing one), but a really naive way of calling bwa from cmdline. Right now if you want to bring in the "correct" branch of BWA, I think messing with @lindenb 's MAKEFILE (bottom lines) and clone only the Apache2 branch would be enough. Per discussion in #1517 , the only difference seem to be `bwa index` performance, which isn't a concern here.

tomwhite commented:
Thanks @SHuang-Broad, I'll give that a go.

BTW you might also be interested in this script for running BWA on YARN: https://github.com/broadinstitute/gatk/commit/4f8a800e0f8b85b8b44c5c70931be3e48aca6bca.

SHuang-Broad commented:
Great, thanks @tomwhite !

tomwhite commented:
@SHuang-Broad @akiezun I've updated the PR to build jbwa, and now the test is passing.

akiezun commented:
@tomwhite - the JNI libraries are now Apache2 licensed (and using Apache2 BWA) and so we are unblocked in using them (courtesy of the author, @lindenb https://github.com/lindenb/jbwa) . Can you add them to this PR so I can test it easily? Thanks

tomwhite commented:
@akiezun our comments overlapped. Let me update again to include the latest Apache 2 JNI libraries.

tomwhite commented:
Updated. Over to you @akiezun.

akiezun commented:
@tomwhite can this take fastq files as input?

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/bwa/BwaSpark.java:
```diff
@@ -0,0 +1,210 @@
+    private static final long serialVersionUID = 1L;
+
+    @Argument(doc = "the reference", shortName = "ref",
+            fullName = "ref", optional = false)
```
we have a special argument for the reference

tomwhite replied:
BWA needs to the load the reference as a file, but the current special argument doesn't support that (`ReferenceSource` just allows you to get the bases, not the file). Something to change in the future perhaps.

akiezun commented:
is jbwa distributed on maven central? if yes, can you use that? If not, let's ask the author. @lindenb is jbwa on central?

Also, all jbwa is now Apache2 so we're good linking to it. No problems there.

lindenb commented:
> is jbwa on central?

@akiezun short answer: no :-)

akiezun commented:
@tomwhite this looks fine to me as a start for us to experiment with but I wish the install was simpler and so we should either include building of jbwa as part of our build or (preferable) use a pre-built version from maven central. wdyt?

back to @tomwhite

akiezun commented:
@lindenb - can you put it up? it would be super

lindenb commented:
@akiezun I've never done this. Do you have any pointer ?

akiezun commented:
@lbergelson can you help?

lbergelson commented:
@akiezun @lindenb I can help. It's probably going to be a slightly complicated process though, especially actually building the cross platform jar.  I assume we're targeting OSX and x86-64 to start with, and then hopefully expanding to POWER8 in the future?  

The general idea is to prebuild the c code for whatever platforms we want to support.  Then package that in a structured way into a jar, and write some java code which will detect the platform at runtime and extract the correct executable into a temporary location.  Then we can publish that jar as a maven artifact.

We have an example of how to do the extraction in the `VectorLoglessPairHMM` constructor.  It's not perfect and probably needs a bit of refactoring to make it more general but it's the right idea.  Other libraries that package native code have similar examples.  I.e. Snappy-java https://github.com/xerial/snappy-java

We're going to be performing similar packaging for other native dependencies that we have, so standardizing it is a good idea.

lindenb commented:
@akiezun @lbergelson FYI: Sorry, I don't have currently the time (and I must say, the need...) to mavenize jbwa on my spare time. Don't wait for me on this please.

akiezun commented:
@lindenb no worries. @lbergelson we do need a solution of either building jbwa in our build script or (better for everyone but more complex short-term) put it on maven.

tomwhite commented:
> can this take fastq files as input?

@akiezun, no it only takes an unaligned BAM as input. This is because it's hard to bring the reads in a pair together without doing a shuffle, when the reads reside in distinct files. However, it's easy to convert from a pair of fastq file to BAM using Picard. E.g.

```
java -jar picard.jar FastqToSam FASTQ=ERR000589_1.1m.fastq FASTQ2=ERR000589_2.1m.fastq OUTPUT=ERR000589.1m.bam SAMPLE_NAME=ERR000589 SORT_ORDER=queryname
```

tomwhite commented:
I ran BWA-MEM on 1m read pairs. Here's a comparison for pure BWA-MEM versus the Spark version, using two threads in each case. (The Spark version had a single executor which was allocated two cores via the `--executor-cores` argument.) 

```
$BWA mem -K 100000 -v 3 -t 2 human_g1k_v37.fasta ERR000589_1.1m.fastq ERR000589_2.1m.fastq > ERR242035.1m.sam 2> foo.err
# Real time: 666.345 sec; CPU: 1267.746 sec

./gatk-launch BwaSpark \
  --threads 2 \
  --ref hdfs://bottou01.sjc.cloudera.com/user/$USER/bwa/human_g1k_v37.fasta \
  --input hdfs:///user/$USER/bwa/ERR000589.1m.bam \
  --output hdfs:///user/$USER/bwa/ERR000589-aligned.1m.bam \
  -- \
  --sparkRunner SPARK \
  --sparkMaster yarn-client \
  --driver-memory 3G \
  --num-executors 1 \
  --executor-cores 2 \
  --executor-memory 15G \
  --archives jbwa-native.tar#jbwa-native \
  --conf 'spark.executor.extraLibraryPath=jbwa-native'
BwaSpark - Total time to run tool: 932s
BwaSpark - Time to download reference: 32s 
BwaSpark - Time in bwa: 1575s
```

So BwaSpark was 40% slower overall in terms of elapsed time. Of course, this was only using a single executor, so if we used multiple executors the elapsed time could be made proportionately shorter (and this a valid use case for some). In terms of CPU time in the BWA algorithm itself, it was about 24% slower (1575/1268). The latter could be improved by optimizing the JNI path (e.g. by passing several arrays of primitives or Strings corresponding to the fields in a `ShortRead`, rather than a single array of `ShortRead`s and then translating it).

There are probably other improvements to make too, but I think the summary is that this approach is feasible.

akiezun commented:
@tomwhite please assign to me when ready for review

tomwhite commented:
@akiezun thanks. I think this is ready for review now. It would be good to merge something that works, even if there are future performance and usability improvements we can do later.

akiezun commented:
I agree, @tomwhite, that there's multiple opportunities for JNI speedup in https://github.com/lindenb/jbwa/blob/master/src/main/native/bwajni.c

akiezun commented:
@tomwhite I did the following to get it to work
1. installed jbwa to `lib` like this (I think that's better than `~`)
```
mkdir lib
cd lib/
mkdir jbwa
git clone https://github.com/lindenb/jbwa   #this uses the apache2 version of bwa so it's all good
cd jbwa/
make
```

and changed build.gradle accordingly

2. add `"-Djava.library.path=lib/jbwa/src/main/native"` to `applicationDefaultJvmArgs` to find the library at runtime (not just in tests)

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/bwa/BwaSpark.java:
```diff
@@ -0,0 +1,328 @@
+
+    @Argument(doc = "the reference", shortName = "ref",
+            fullName = "ref", optional = false)
+    private String ref;
```
use `StandardArgumentDefinitions.REFERENCE_LONG_NAME` and StandardArgumentDefinitions.REFERENCE_SHORT_NAME` for the reference argument

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/bwa/BwaSpark.java:
```diff
@@ -0,0 +1,328 @@
+    @Override
+    protected void runTool(final JavaSparkContext ctx) {
+        Stopwatch toolStopwatch = Stopwatch.createStarted();
+        ctx.hadoopConfiguration().setBoolean(BAMInputFormat.KEEP_PAIRED_READS_TOGETHER_PROPERTY, true);
```
what reads this property? Can you add a comment?

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/bwa/BwaSpark.java:
```diff
@@ -0,0 +1,328 @@
+        Stopwatch toolStopwatch = Stopwatch.createStarted();
+        ctx.hadoopConfiguration().setBoolean(BAMInputFormat.KEEP_PAIRED_READS_TOGETHER_PROPERTY, true);
+        JavaRDD<GATKRead> unalignedReads = getReads();
+        JavaRDD<List<GATKRead>> unalignedPairs = unalignedReads.mapPartitions(iter -> () -> pairwise(iter));
```
should this call to `pairwise` just be inlined? It may make it simpler to read I think

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/bwa/BwaSpark.java:
```diff
@@ -0,0 +1,328 @@
+                    new ShortRead(name2, read2.getBases(), baseQualities2));
+        });
+
+        JavaRDD<String> samLines = shortReadPairs.mapPartitions(iter -> () -> {
```
can you extract everything up to `samLines` into one method and then the rest into another method? Ie, make it 3 clear steps: read, align, write

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/bwa/BwaSpark.java:
```diff
@@ -0,0 +1,328 @@
+        }
+    }
+
+    static <T> Iterator<T> concat(Iterator<? extends Iterable<T>> iterator) {
```
can you add a comment what this method does? Also, make it private or add tests for it

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/bwa/BwaSpark.java:
```diff
@@ -0,0 +1,328 @@
+        };
+    }
+
+    private Iterator<List<String>> alignChunks(final BwaMem bwaMem, Iterator<Tuple2<ShortRead, ShortRead>> iter, int fixedChunkSize) {
```
add 1 sentence what this method does

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/bwa/BwaSpark.java:
```diff
@@ -0,0 +1,328 @@
+        }, numThreads);
+    }
+
+    private <U> Iterator<List<Tuple2<ShortRead, ShortRead>>> chunk(final Iterator<Tuple2<ShortRead, ShortRead>> iterator, int fixedChunkSize) {
```
add a short comment for this method - what it does

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/bwa/BwaSpark.java:
```diff
@@ -0,0 +1,328 @@
+
+    public int NUM_SEQUENCES = Integer.MAX_VALUE;
+
+    SAMSequenceDictionary makeSequenceDictionary(final File referenceFile) {
```
make private

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/bwa/BwaSpark.java:
```diff
@@ -0,0 +1,328 @@
+        return makeSequenceDictionary(referenceFile.toPath());
+    }
+
+    SAMSequenceDictionary makeSequenceDictionary(final Path referenceFile) {
```
private

akiezun replied:
can you delegate to `CreateSequenceDictionary.makeSequenceDictionary` or extract a method for both to use? @droazen is there a way to simply directly load a `.dict` file?

akiezun replied:
@droazen just told me how to do it - use `new IndexedFastaSequenceFile(file).getSequenceDictionary()` (it will return null if not there and they you can use `CreateSequenceDictionary.makeSequenceDictionary`. Let's not have that copied code here.

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/bwa/BwaSpark.java:
```diff
@@ -0,0 +1,328 @@
+    }
+
+    /**
+     * Create one SAMSequenceRecord from a single fasta sequence
```
let's either get rid of this code (by reading a dict directly) or at least document that it's copied from `CreateSequenceDictionary.makeSequenceRecord`

akiezun commented:
can you add a test for the actual BWA alinment? Ie take some reads and align them and confirm that they get aligned properly (ie same way as if bwa was called on commandline)?

akiezun commented:
back to @tomwhite - some small refactor, building and testing suggestions.

sputnikci commented on src/test/java/org/broadinstitute/hellbender/tools/spark/bwa/BwaSparkIntegrationTest.java:
```diff
@@ -0,0 +1,37 @@
+        final File ref = getTestFile("ref.fa");
+        final File input = getTestFile("R.bam");
+        final File output = createTempFile("bwa", ".bam");
+        output.delete();
```
\[FindBugs\] INFO: RV: Exceptional return value of java\.io\.File\.delete\(\) ignored in org\.broadinstitute\.hellbender\.tools\.spark\.bwa\.BwaSparkIntegrationTest\.test\(\)

sputnikci replied:
\[FindBugs\] INFO: RV: Exceptional return value of java\.io\.File\.delete\(\) ignored in org\.broadinstitute\.hellbender\.tools\.spark\.bwa\.BwaSparkIntegrationTest\.test\(\)

sputnikci replied:
\[FindBugs\] INFO: RV: Exceptional return value of java\.io\.File\.delete\(\) ignored in org\.broadinstitute\.hellbender\.tools\.spark\.bwa\.BwaSparkIntegrationTest\.test\(\)

sputnikci replied:
\[FindBugs\] INFO: RV: Exceptional return value of java\.io\.File\.delete\(\) ignored in org\.broadinstitute\.hellbender\.tools\.spark\.bwa\.BwaSparkIntegrationTest\.test\(\)

tomwhite commented:
Thanks for the review @akiezun. I've addressed all your points. I dropped the sequence dictionary code, so now we rely on a .dict file being present (and if it isn't you can generate one using Picard's CreateSequenceDictionary command).

I think the only outstanding issue that is stopping this from being merged is integrating the build of jbwa. Currently you have to manually checkout and build jbwa yourself. The travis build does this automatically, but the local gradle build doesn't. Any thoughts about how to address this?

tomwhite commented:
The latest command line I used for testing:

```
./gatk-launch BwaSpark \
  --threads 2 \
  --reference hdfs://bottou01.sjc.cloudera.com/user/$USER/bwa/human_g1k_v37.fasta \
  --input hdfs:///user/$USER/bwa/ERR000589.100k.bam \
  --output hdfs:///user/$USER/bwa/ERR000589-aligned.100k.bam \
  --disableSequenceDictionaryValidation true \
  -- \
  --sparkRunner SPARK \
  --sparkMaster yarn-client \
  --driver-memory 3G \
  --num-executors 1 \
  --executor-cores 2 \
  --executor-memory 15G \
  --archives lib/jbwa/jbwa-native.tar#jbwa-native \
  --conf 'spark.executor.extraLibraryPath=jbwa-native'
```

akiezun commented:
ready for rereview?

tomwhite commented:
Yes, please take a look @akiezun.

akiezun commented:
@tomwhite about the building - ideally jbwa would be on maven central and have so and jnilib built in. Until that happens we can only
A) expect people to have it installed (seems bad), or
B) include script to install it, or
C) include pre-built binaries (.so, .jnilib and .jar)

I vote for B so that no steps are magic - (on the mac https://github.com/lindenb/jbwa/pull/5 was required but it's now merged). You can call the install script from gradle like this and make `installDist` depend on `installJBWA`

```
task installJBWA(type: Exec) {
  executable "sh"
  args "-x", "installJBWA.sh"
}
```

my  `installJBWA.sh` is 
```
mkdir -p lib
cd lib/
mkdir -p jbwa
git clone https://github.com/lindenb/jbwa
cd jbwa/
make
#Note: this rename is only needed on the Mac but does not cause problems elsewhere
cp -v src/main/native/libbwajni.so src/main/native/libbwajni.jnilib
``` 
wdyt?

akiezun commented:
back to @tomwhite to include the install script and we can merge it

droazen commented:
@akiezun I don't think we should add yet another Frankenstein monster step to our build process like this. I recommend instead that we package the library in Broad artifactory (after obtaining the author's permission), and depend on it from there.

akiezun commented:
do you know how?

On Wed, May 18, 2016 at 3:50 PM, droazen <notifications@github.com> wrote:

> @akiezun <https://github.com/akiezun> I don't think we should add yet
> another Frankenstein monster step to our build process like this. I
> recommend instead that we package the library in Broad artifactory (after
> obtaining the author's permission), and depend on it from there.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/broadinstitute/gatk/pull/1750#issuecomment-220138064>
>

droazen commented:
I'm sure we can figure it out based on the code we already have in build.gradle to upload gatk snapshots to the artifactory.

This seems much better than some gross shell script and more pain and suffering for people trying to build GATK4.

akiezun commented:
script is not pretty but there's no pain. It just builds it. And it solves
the problem of mac vs linux. In your scenario, how does that work?

On Wed, May 18, 2016 at 4:12 PM, droazen <notifications@github.com> wrote:

> I'm sure we can figure it out based on the code we already have in
> build.gradle to upload gatk snapshots to the artifactory.
>
> This seems much better than some gross shell script and more pain and
> suffering for people trying to build GATK4.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/broadinstitute/gatk/pull/1750#issuecomment-220143766>
>

droazen commented:
We just package both Mac and Linux versions of the library in a jar on the artifactory. Problem solved, no pain for users.

Your solution could potentially cause user pain if the native build fails due to missing dependencies (don't say it won't happen -- you said that about OpenMP and it happened yesterday!). It also doesn't address the packaging issue at all.

akiezun commented:
so you don't think https://github.com/broadinstitute/gatk/blob/master/scripts/upload_artifact_with_both_lib_and_dylib.sh is nasty and painful? 
(hint: it ssh's to gsa6 and builds and fetches files back! it requires that the user be on Broad network)

droazen commented:
@akiezun The difference is that `upload_artifact_with_both_lib_and_dylib.sh` is only run by us when we release -- it does not affect builds by users of the GATK, like your approach would.

We need to stop adding garbage to the GATK4 build system before it's too late and ends up like the GATK3 build system. We want GATK4 to be easy and reliable to build from source.

droazen commented:
Also, we've already agreed as a team that packaging pre-built native libraries for the platforms and OSes that we support into jars and depending on them via maven is the right thing to do for native dependencies of the GATK. We can do this right now for jbwa via the artifactory, with minimal additional effort on our part -- so why are you defending the "hack build.gradle some more and make GATK even harder to reliably build" approach to native code that we've already agreed is bad?

akiezun commented:
honestly, I really don't care which way we pick as long as it works. Can you make it work?

droazen commented:
I can help with the "publishing to artifactory" part if you provide me with pre-built Mac and Linux versions of the library and obtain permission from the author.

akiezun commented:
for the libraries: run my script from a few comments above. @lindenb can we put a pre built versions of this library on https://artifactory.broadinstitute.org/artifactory/webapp/home.html until a maven central artifact is created ?

lindenb commented:
>  @lindenb can we put a pre built versions of this library on https://artifactory.broadinstitute.org/artifactory/webapp/home.html until a maven central artifact is created ? 

yes, no problem.

droazen commented:
Alright, I'll try to get it published to the Broad artifactory by end of the week. Once it's there, this branch will need to be modified to depend on the version in artifactory, locate the library on the classpath, and extract and load it.

droazen commented:
This PR is waiting on https://github.com/broadinstitute/gatk/issues/1838

droazen commented:
PR resolving https://github.com/broadinstitute/gatk/issues/1838 has been opened:  https://github.com/broadinstitute/gatk/pull/1847. I included a method to extract and load the library from the classpath, as well as a unit test showing how to select between the Linux and Mac versions of the library.

droazen commented:
@tomwhite #1847 is in, and there's now a `JBWAIntegrationTest` showing how to load and use jbwa in a way that will work on both Linux and Mac. You should be all set -- let me know if there are problems.
